browser/src/control/jsdialog/Widget.TreeView.js

```diff
@@ -162,19 +164,25 @@ function _treelistboxEntry(parentContainer, treeViewData, entry, builder, isTree
 	}
 
 	var toggleFunction = function() {
+		if (L.DomUtil.hasClass(span, 'collapsed'))
+			builder.callback('treeview', 'expand', treeViewData, entry.row, builder);
+		else
+			builder.callback('treeview', 'collapse', treeViewData, entry.row, builder);
 		$(span).toggleClass('collapsed');
 	};
 
 	var expandFunction = function () {
 		if (entry.ondemand && L.DomUtil.hasClass(span, 'collapsed'))
 			builder.callback('treeview', 'expand', treeViewData, entry.row, builder);
-		toggleFunction();
+		$(span).toggleClass('collapsed');
 	};
 
 	if (entry.children) {
 		var ul = L.DomUtil.create('ul', builder.options.cssClass, li);
-		for (var i in entry.children) {
-			_treelistboxEntry(ul, treeViewData, entry.children[i], builder, isTreeView, treeRoot);
+		if (!entry.collapsed) {
```

Comment:
we should create even collapsed children. we use treeview also without dialog in the backend (in Zotero plugin) where we just show/hide nodes on collapse/expand, there is no message from the core after change